### DISCUSSION TITLE
can AssignAction be returned form the actions hash of the second argument to createMachine

### DISCUSSION THREAD

**Discussion by dagda1** [statelyai/xstate#3049] [Q&A]
I've just recently upgraded to xstate `0.3.0` and the language server is highlighting all inline actions like this:

```ts
loading: {
  on: {
    SUCCESS: [{
      target: 'succeeded',
      actions: model.assign(
        {
          data: (_, e) => e.payload,
          error: undefined
        },
        'SUCCESS'
      ),
    }]
  }
},
```
The following warning is displayed
> Inline implementations cannot be used with the visual editor

I was having a look at the examples directory and the [react-tic-tac-toe](https://github.com/statelyai/xstate/blob/main/examples/react-tic-tac-toe/src/ticTacToe.machine.ts#L113) example has this examples hash:

```ts
    actions: {
      // @ts-ignore (will be fixed in next minor version)
      updateBoard: model.assign(
        {
          board: (context, event) => {
            const updatedBoard = [...context.board];
            updatedBoard[event.value] = context.player;
            return updatedBoard;
          },
          moves: (context) => context.moves + 1,
          player: (context) => (context.player === 'x' ? 'o' : 'x')
        },
        'PLAY'
      ),
```
There is a `@ts-ignore` above `updateBoard` as it does not return an `ActionObject` type.

The comment is 6 months old.

It is confusing as to whether this is a good way of doing things.

**Comment by davidkpiano** [comment]
This will be fixed soon (cc. @mattpocock)

**Comment by Andarist** [comment]
Note that this warning should go away because Matt is working on preserving inline implementations while editing a machine in the editor.

However, the `@ts-ignore` stuff is a different issue - `model.assign` actually returns an `ActionObject` so this part is OK. However, this particular `ActionObject` shouldn't be assignable to this position! This one was declared as accepting only the PLAY event but we can't guarantee that at the type-level here. I've dug into this quite a bit and it turned out that this is an issue in TypeScript itself: https://github.com/microsoft/TypeScript/issues/44572#issuecomment-1042915921 

I've prepared a PR refactoring this particular example to use a more sound approach: https://github.com/statelyai/xstate/pull/3060 . Our general recommendation now is to use the recently released typegen because it's the only solution that can deliver you great types without friction.
